<!DOCTYPE html>
<html lang="zh-CN">
<head>
  <meta charset="UTF-8">
  <meta http-equiv="X-UA-Compatible" content="IE=edge">
  <meta name="viewport" content="width=device-width, initial-scale=1.0">
  <title>Promise_A</title>
</head>
<body>
  <script>
    // 期约基础

    // ECMAScript 6 新增的引用类型 Promise，可以通过 new 操作符来实例化。
    // 创建新期约时需要传入执行器（executor）函数作为参数。
    let p = new Promise(() => {});
    setTimeout(console.log, 0, p); // Promise {<pending>}

    // 1. 期约状态机
    // 在把一个期约实例传给 console.log() 时，控制台输出（可能因浏览器不同而略有差异）表明该实例
    // 处于待定（pending）状态。如前所述，期约是一个有状态的对象，可能处于如下 3 种状态之一：
    // 待定（pending）；兑现（fulfilled，有时候也称为“解决”，resolved）；拒绝（rejected）

    // 待定（pending）是期约的最初始状态。在待定状态下，期约可以落定（settled）为代表成功的兑现
    // （fulfilled）状态，或者代表失败的拒绝（rejected）状态。无论落定为哪种状态都是不可逆的。
    // 只要从待定转换为兑现或拒绝，期约的状态就不再改变。而且，也不能保证期约必然会脱离待定状态。
    // 因此，组织合理的代码无论期约解决（resolve）还是拒绝（reject），甚至永远处于待定（pending）状态，
    // 都应该具有恰当的行为。

    // 重要的是，期约的状态是私有的，不能直接通过 JavaScript 检测到。这主要是为了避免根据
    // 读取到的期约状态，以同步方式处理期约对象。另外，期约的状态也不能被外部 JavaScript 代码修改。
    // 这与不能读取该状态的原因是一样的：期约故意将异步行为封装起来，从而隔离外部的同步代码。

    // 2. 解决值、拒绝理由及期约用例
    // 期约主要有两大用途。首先是抽象地表示一个异步操作。期约的状态代表期约是否完成。
    // “待定”表示尚未开始或者正在执行中。“兑现”表示已经成功完成，而“拒绝”则表示没有成功完成。

    // 某些情况下，这个状态机就是期约可以提供的最有用的信息。知道一段异步代码已经完成，对于其他代码
    // 而言已经足够了。比如，假设期约要向服务器发送一个 HTTP 请求。请求返回 200~299 范围内的状态码
    // 就足以让期约的状态变为“兑现”。类似地，如果请求返回的状态码不在 200~299 这个范围内，
    // 那么就会把期约状态切换为“拒绝”。

    // 在另外一些情况下，期约封装的异步操作会实际生成某个值，而程序期待期约状态改变时可以访问这个值。
    // 相应地，如果期约被拒绝，程序就会期待期约状态改变时可以拿到拒绝的理由。比如，假设期约向服务器
    // 发送一个 HTTP 请求并预定会返回一个 JSON。如果请求返回范围在 200~299 的状态码，则足以让
    // 期约的状态变为兑现。此时期约内部就可以收到一个 JSON 字符串。类似地，如果请求返回的状态码
    // 不在 200~299 这个范围内，那么就会把期约状态切换为拒绝。此时拒绝的理由可能是一个 Error 对象，
    // 包含着 HTTP 状态码及相关错误消息。

    // 为了支持这两种用例，每个期约只要状态切换为兑现，就会有一个私有的内部值（value）。
    // 类似地，每个期约只要状态切换为拒绝，就会有一个私有的内部理由（reason）。无论是值还是理由，
    // 都是包含原始值或对象的不可修改的引用。二者都是可选的，而且默认值为 undefined。
    // 在期约到达某个落定状态时执行的异步代码始终会收到这个值或理由。

    // 3. 通过执行函数控制期约状态
    // 由于期约的状态是私有的，所以只能在内部进行操作。内部操作在期约的执行器函数中完成。
    // 执行器函数主要有两项职责：初始化期约的异步行为和控制状态的最终转换。其中，控制期约状态的转换
    // 是通过调用它的两个函数参数实现的。这两个函数参数通常都命名为 resolve() 和 reject()。
    // 调用 resolve() 会把状态切换为兑现，调用 reject() 会把状态切换为拒绝。调用 reject() 也会抛出错误。
    let p1 = new Promise((resolve, reject) => resolve()); 
    setTimeout(console.log, 0, p1); // Promise {<fulfilled>: undefined}
    let p2 = new Promise((resolve, reject) => reject());
    setTimeout(console.log, 0, p2); // Promise {<rejected>: undefined}

    // 在上面的例子中，并没有什么异步操作，因为在初始化期约时，执行器函数已经改变了每个期约的状态。
    // 这里的关键在于，执行器函数是同步执行的。这是因为执行器函数是期约的初始化程序。
    new Promise(() => setTimeout(console.log, 0, 'executor'));
    setTimeout(console.log, 0, 'promise initialized');
    // executor
    // promise initialized

    // 添加 setTimeout 可以推迟切换状态。
    let p3 = new Promise((resolve, reject) => setTimeout(resolve, 1000)); // 1 秒后调用 reject() 
    // 在 console.log 打印期约实例的时候，还不会执行超时回调（即 resolve()）
    setTimeout(console.log, 0, p3);           // Promise {<pending>}
    setTimeout(console.log, 0, p3); // 1 秒后 // Promise {<fulfilled>: undefined}
    
    // 无论 resolve() 和 reject()中的哪个被调用，状态转换都不可撤销了。于是继续修改状态会静默失败。
    let p4 = new Promise((resolve, reject) => { 
     resolve();
     reject(); // 没有效果
    }); 
    setTimeout(console.log, 0, p4);
    // Promise {<fulfilled>: undefined}

    // 为避免期约卡在待定状态，可以添加一个定时退出功能。
    let p5 = new Promise((resolve, reject) => setTimeout(reject, 10000)); // 10 秒后调用 reject() 
    setTimeout(console.log, 0, p5);                // Promise {<pending>}
    setTimeout(console.log, 11000, p5); // 11 秒后 // Promise {<rejected>: undefined}

    // 因为期约的状态只能改变一次，所以这里的超时拒绝逻辑中可以放心地设置让期约处于待定状态的最长时间。
    // 如果执行器中的代码在超时之前已经解决或拒绝，那么超时回调再尝试拒绝也会静默失败。

    // 4. Promise.resolve();
    // 期约并非一开始就必须处于待定状态，然后通过执行器函数才能转换为落定状态。通过调用
    // Promise.resolve() 静态方法，可以实例化一个解决的期约。下面两个期约实例实际上是一样的：
    let pp1 = new Promise((resolve, reject) => resolve()); 
    let pp2 = Promise.resolve();

    // 这个解决的期约的值对应着传给 Promise.resolve() 的第一个参数。
    // 使用这个静态方法，实际上可以把任何值都转换为一个期约。
    setTimeout(console.log, 0, Promise.resolve());        // Promise {<fulfilled>: undefined}
    setTimeout(console.log, 0, Promise.resolve(3));       // Promise {<fulfilled>: 3}
    setTimeout(console.log, 0, Promise.resolve(4, 5, 6)); // Promise {<fulfilled>: 4}（多余的参数会忽略）

    // 对这个静态方法而言，如果传入的参数本身是一个期约，那它的行为就类似于一个空包装。
    // Promise.resolve() 可以说是一个幂等方法。
    let pp3 = Promise.resolve(7); 
    setTimeout(console.log, 0, pp3 === Promise.resolve(pp3));                  // true
    setTimeout(console.log, 0, pp3 === Promise.resolve(Promise.resolve(pp3))); // true

    // 这个幂等性会保留传入期约的状态：
    let pp4 = new Promise(() => {}); 
    setTimeout(console.log, 0, pp4);                          // Promise {<pending>}
    setTimeout(console.log, 0, Promise.resolve(pp4));         // Promise {<pending>}
    setTimeout(console.log, 0, pp4 === Promise.resolve(pp4)); // true

    // 这个静态方法能够包装任何非期约值，包括错误对象，并将其转换为解决的期约。也可能导致不符合预期的行为。
    let pp5 = Promise.resolve(new Error('foo')); 
    setTimeout(console.log, 0, pp5); // Promise {<fulfilled>: Error: foo}

    // 5. Promise.reject()
    // Promise.reject() 与 Promise.resolve() 类似，会实例化一个拒绝的期约并抛出一个异步错误
    // （这个错误不能通过 try/catch 捕获，而只能通过拒绝处理程序捕获）。下面的两个期约实例实际是一样的：
    let ppp1 = new Promise((resolve, reject) => reject()); // Uncaught (in promise) undefined
    let ppp2 = Promise.reject();                           // Uncaught (in promise) undefined

    // 这个拒绝的期约的理由就是传给 Promise.reject() 的第一个参数。这个参数也会传给后续的拒绝处理程序。 
    let ppp3 = Promise.reject(3); 
    setTimeout(console.log, 0, ppp3);                      // Promise {<rejected>: 3}
    ppp3.then(null, (e) => setTimeout(console.log, 0, e)); // 3

    // 关键在于，Promise.reject()并没有照搬 Promise.resolve()的幂等逻辑。
    // 如果给它传一个期约对象，则这个期约会成为它返回的拒绝期约的理由。
    setTimeout(console.log, 0, Promise.reject(Promise.resolve())); // Promise {<rejected>: Promise}
</script>
</body>
</html>